src/protocols/secure_channel/CheckinMessage.cpp

```diff
@@ -0,0 +1,110 @@
+
+static constexpr uint16_t sMinPayloadSize =
+    CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES + sizeof(uint32_t) + CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES;
+
+CHIP_ERROR CheckinMessage::GenerateCheckinMessagePayload(Crypto::Aes128KeyHandle & key, uint32_t counter, const ByteSpan & appData,
+                                                         MutableByteSpan & output, uint16_t * payloadSize)
+{
+    VerifyOrReturnError(appData.size() <= CHIP_CHECK_IN_APP_DATA_MAX_SIZE, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(output.size() >= (appData.size() + sMinPayloadSize), CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err            = CHIP_NO_ERROR;
+    uint8_t * appDataStartPtr = output.data() + CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES;
+
+    // Write32 increase the pointer by sizeof(uint32_t)
+    Encoding::LittleEndian::Write32(appDataStartPtr, counter);
+    appDataStartPtr -= sizeof(uint32_t);
+
+    {
+        chip::Crypto::HMAC_sha shaHandler;
+        uint8_t nonceWorkBuffer[CHIP_CRYPTO_HASH_LEN_BYTES] = { 0 };
+        ReturnErrorOnFailure(shaHandler.HMAC_SHA256(key.As<Aes128KeyByteArray>(), sizeof(Aes128KeyByteArray), appDataStartPtr,
+                                                    sizeof(counter), nonceWorkBuffer, CHIP_CRYPTO_HASH_LEN_BYTES));
+        memcpy(output.data(), nonceWorkBuffer, CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES);
+    }
+
+    // In place encryption to save some RAM
+    memcpy(appDataStartPtr + sizeof(counter), appData.data(), appData.size());
+
+    // API constraint. Mic buffer needs to be separated.
+    uint8_t mic[CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES];
+    uint8_t * micPtr = appDataStartPtr + appData.size() + sizeof(counter);
+
+    ReturnErrorOnFailure(Crypto::AES_CCM_encrypt(appDataStartPtr, appData.size() + sizeof(counter), nullptr, 0, key, output.data(),
+                                                 CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES, appDataStartPtr, mic,
+                                                 CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES));
+
+    memcpy(micPtr, mic, CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES);
+
+    if (payloadSize != nullptr)
+    {
+        *payloadSize = static_cast<uint16_t>(appData.size() + sMinPayloadSize);
+    }
+
+    return err;
+}
+
+CHIP_ERROR CheckinMessage::ParseCheckinMessagePayload(Crypto::Aes128KeyHandle & key, ByteSpan & payload, uint32_t & counter,
+                                                      MutableByteSpan & appData)
+{
+    VerifyOrReturnError(payload.size() >= sMinPayloadSize, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(payload.size() <= (sMinPayloadSize + CHIP_CHECK_IN_APP_DATA_MAX_SIZE), CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    uint16_t appDataSize = GetAppDataSize(payload);
+
+    // To prevent workbuffer usage, appData size need to be large enough to hold both the appData and the counter
+    VerifyOrReturnError(appData.size() >= appDataSize + sizeof(uint32_t), CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
Again, in all of these additions making the order of the terms match the data order would be clearer...

And maybe we should just have a `constexpr size_t counterSize = sizeof(uint32_t)` to make that part really clear?  Then maybe we don't need the CounterType thing.